### PR TITLE
feat(Auth): Add TOTP Support (#2537)

### PR DESCRIPTION
**Pull request by gpanshu**
- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:*

*How did you test these changes?*
(Please add a line here how the changes were tested)

*Documentation update required?*
- [ ] No
- [y] Yes [(Please include a PR link for the documentation update)
](https://github.com/aws-amplify/docs/pull/5768)
*General Checklist*
- [x] Added Unit Tests
- [ ] Added Integration Tests
- [x] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by ankpshah — APPROVED** [review]

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/exceptions/service/EnableSoftwareTokenMfaException.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ */
+package com.amplifyframework.auth.cognito.exceptions.service
+
+import com.amplifyframework.auth.AuthException
+
+/**
+ * Software Token MFA is not enabled for the user.
+ * @param cause The underlying cause of this exception
+ */
+open class EnableSoftwareTokenMfaException(cause: Throwable?) :
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/exceptions/service/EnableSoftwareTokenMfaException.kt:23]: Doc says `EnableSoftwareTokenMFAException`.

**Review by tylerjroach — CHANGES_REQUESTED** [review]

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/KotlinAuthFacadeInternal.kt [hunk 2]
```diff
@@ -520,4 +522,46 @@ internal class KotlinAuthFacadeInternal(private val delegate: RealAWSCognitoAuth
             )
         }
     }
+
+    suspend fun setupMFA(): TOTPSetupDetails {
```
[the remark this answers is not in the document]
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/KotlinAuthFacadeInternal.kt:526]: done

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/options/AWSCognitoAuthVerifyTOTPSetupOptions.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        return friendlyDeviceName;
+    }
+
+    /**
+     * The builder for this class.
+     */
+    public static final class CognitoBuilder extends Builder<CognitoBuilder> {
+        private String friendlyDeviceName;
+
+        private String getFriendlyDeviceName() {
```
[the remark this answers is not in the document]
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/options/AWSCognitoAuthVerifyTOTPSetupOptions.java:45]: done

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/options/AWSCognitoAuthVerifyTOTPSetupOptions.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+ */
+
+package com.amplifyframework.auth.cognito.options;
+
+import com.amplifyframework.auth.options.AuthVerifyTOTPSetupOptions;
+
+/**
+ * Cognito extension of update verify totp setup options to add the platform specific fields.
+ */
+public final class AWSCognitoAuthVerifyTOTPSetupOptions extends AuthVerifyTOTPSetupOptions {
```
[the remark this answers is not in the document]
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/options/AWSCognitoAuthVerifyTOTPSetupOptions.java:23]: done

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/options/AWSCognitoAuthVerifyTOTPSetupOptions.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        private String getFriendlyDeviceName() {
+            return friendlyDeviceName;
+        }
+
+        /**
+         * Friendly device name to be set in Cognito.
+         * @param friendlyDeviceName String input for friendlyDeviceName
+         * @return current CognitoBuilder instance
+         * */
+        public CognitoBuilder setFriendlyDeviceName(String friendlyDeviceName) {
```
[the remark this answers is not in the document]
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/options/AWSCognitoAuthVerifyTOTPSetupOptions.java:54]: done

File: core-kotlin/src/main/java/com/amplifyframework/kotlin/auth/Auth.kt [hunk 6]
```diff
@@ -327,4 +329,20 @@ interface Auth {
      */
     @Throws(AuthException::class)
     suspend fun deleteUser()
+
+    /**
+     * Setup TOTP for the currently signed in user.
+     * @return TOTP Setup details
+     */
+    suspend fun setUpTOTP(): TOTPSetupDetails
```
[the remark this answers is not in the document]
Comment (gpanshu) [reply core-kotlin/src/main/java/com/amplifyframework/kotlin/auth/Auth.kt:337]: Done

File: core-kotlin/src/main/java/com/amplifyframework/kotlin/auth/Auth.kt [hunk 7]
```diff
@@ -327,4 +329,20 @@ interface Auth {
...
+
+    /**
+     * Verify TOTP setup for the currently signed in user.
+     * @param code TOTP code to verify TOTP setup
+     * @param options additional options to verify totp setup
+     */
+    suspend fun verifyTOTPSetup(
```
[the remark this answers is not in the document]
Comment (gpanshu) [reply core-kotlin/src/main/java/com/amplifyframework/kotlin/auth/Auth.kt:344]: done

File: core/src/main/java/com/amplifyframework/auth/MFAType.kt [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amplifyframework.auth
+
+enum class MFAType(val value: String) {
+    SMS("SMS_MFA"),
+    TOTP("SOFTWARE_TOKEN_MFA");
+    companion object {
+        fun toMFAType(value: String): MFAType {
```
[the remark this answers is not in the document]
Comment (gpanshu) [reply core/src/main/java/com/amplifyframework/auth/MFAType.kt:21]: Its sent back to the customer as part of AuthSignInStep so we need it to be public

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/options/AWSCognitoAuthVerifyTOTPSetupOptions.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tylerjroach) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/options/AWSCognitoAuthVerifyTOTPSetupOptions.java]: Sorry, just to be clear Cognito should be in name as it is on all other Options. The review doc needs to be corrected.

File: core/src/main/java/com/amplifyframework/auth/MFAType.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (tylerjroach) [reply core/src/main/java/com/amplifyframework/auth/MFAType.kt:21]: The comment was on the `toMFAType` function, not on `MFAType`

File: aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginTOTPTests.kt [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+    private lateinit var authPlugin: AWSCognitoAuthPlugin
+    private lateinit var synchronousAuth: SynchronousAuth
+    private val password = UUID.randomUUID().toString()
+    private val userName = "testUser${Random().nextInt()}"
+    private val email = "$userName@testdomain.com"
+
+    @Before
+    fun setup() {
+        val context = ApplicationProvider.getApplicationContext<Context>()
+        Amplify.addPlugin(AndroidLoggingPlugin(LogLevel.VERBOSE))
```
Comment (tylerjroach) [inline aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginTOTPTests.kt:57]: This shouldn't need to be here.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 10]
```diff
@@ -1986,6 +2113,200 @@ internal class RealAWSCognitoAuthPlugin(
         }
     }
 
+    override fun setUpTOTP(onSuccess: Consumer<TOTPSetupDetails>, onError: Consumer<AuthException>) {
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:2116]: Is there no state machine for this block?

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 11]
```diff
@@ -1986,6 +2113,200 @@ internal class RealAWSCognitoAuthPlugin(
...
+                        }
+                    }
+                }
+
+                else -> onError.accept(InvalidStateException())
+            }
+        }
+    }
+
+    fun updateMFAPreference(
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:2219]: No state machine?

File: aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginTest.kt [hunk 12]
```diff
@@ -681,6 +683,53 @@ class AWSCognitoAuthPluginTest {
         verify(timeout = CHANNEL_TIMEOUT) { realPlugin.clearFederationToIdentityPool(any(), any()) }
     }
 
+    @Test
+    fun setUpTOTP() {
```
Comment (tylerjroach) [inline aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginTest.kt:687]: For this test and the 3 tests below it, `any()` shouldn't be used on the last blocks. Its should be asserting that the correct listeners, etc are passed. In this case it should check expectedOnSuccess and expectedOnError.

**Review by tylerjroach — CHANGES_REQUESTED** [review]
Feature tests are missing

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/UserMFAPreference.kt [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amplifyframework.auth.cognito
+
+import com.amplifyframework.auth.MFAType
+
+public data class UserMFAPreference(
```
[the remark this answers is not in the document]
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/UserMFAPreference.kt:19]: Lets sync offline cause it does look like it matches

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/exceptions/service/EnableSoftwareTokenMfaException.kt (hunk 1 above)
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/exceptions/service/EnableSoftwareTokenMfaException.kt]: done

File: aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginTOTPTests.kt (hunk 9 above)
Comment (gpanshu) [reply aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginTOTPTests.kt:57]: removed

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt (hunk 11 above)
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:2219]: yeah its not required as that will be extra overhead. The api is called directly.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt (hunk 10 above)
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:2116]: yeah its not required as that will be extra overhead. The api is called directly.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/options/AWSCognitoAuthVerifyTOTPSetupOptions.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/options/AWSCognitoAuthVerifyTOTPSetupOptions.java]: Thanks for clearing that up. Updated.

File: core/src/main/java/com/amplifyframework/TOTPSetupDetails.kt [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+package com.amplifyframework
+
+import android.net.Uri
+
+data class TOTPSetupDetails(
+    val sharedSecret: String,
+    val username: String
+) {
+    fun getSetupURI(
+        appName: String,
```
[the remark this answers is not in the document]
Comment (gpanshu) [reply core/src/main/java/com/amplifyframework/TOTPSetupDetails.kt]: done

File: core/src/main/java/com/amplifyframework/auth/MFAType.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (gpanshu) [reply core/src/main/java/com/amplifyframework/auth/MFAType.kt:21]: Its used in RealAwsPlugin and hence can not be marked `internal`. I can update the api doc.

**Review by tylerjroach — DISMISSED** [review]

**Review by mattcreaser — APPROVED** [review]

**Review by tylerjroach — APPROVED** [review]

**Pull request merged**